File: 3129-find-all-possible-stable-binary-arrays-i/3129-find-all-possible-stable-binary-arrays-i.cpp
class Solution {
public:
    int numberOfStableArrays(int zero, int one, int limit) {
        long long MOD = 1000000007;
        int maxN = zero + one;
        
        std::vector<long long> fact(maxN + 1, 0);
        std::vector<long long> invFact(maxN + 1, 0);
        
        fact[0] = 1;
        invFact[0] = 1;
        for (int i = 1; i <= maxN; i++) {
            fact[i] = (fact[i - 1] * i) % MOD;
        }
        
        auto power = [&](long long baseVal, long long exp) {
            long long res = 1;
            baseVal %= MOD;
            while (exp > 0) {
                if (exp & 1) res = (res * baseVal) % MOD;
                baseVal = (baseVal * baseVal) % MOD;
                exp >>= 1;
            }
            return res;
        };
        
        invFact[maxN] = power(fact[maxN], MOD - 2);
        for (int i = maxN - 1; i >= 1; i--) {
            invFact[i] = (invFact[i + 1] * (i + 1)) % MOD;
        }
        
        auto C = [&](int n, int k) -> long long {
            if (k < 0 || k > n) return 0;
            return fact[n] * invFact[k] % MOD * invFact[n - k] % MOD;
        };
        
        auto F = [&](int N, int K, int L) -> long long {
            if (K <= 0 || K > N) return 0;
            long long ans = 0;
            int maxJ = (N - K) / L;
            for (int j = 0; j <= maxJ; j++) {
                long long term = C(K, j) * C(N - j * L - 1, K - 1) % MOD;
                if (j & 1) {
                    ans = (ans - term + MOD) % MOD;
                } else {
                    ans = (ans + term) % MOD;
                }
            }
            return ans;
        };

        int maxK = std::min(zero, one + 1);
        std::vector<long long> fOne(maxK + 2, 0);
        for (int k = 1; k <= maxK + 1; k++) {
            fOne[k] = F(one, k, limit);
        }
        
        long long ans = 0;
        for (int k = 1; k <= maxK; k++) {
            long long fz = F(zero, k, limit);
            if (fz == 0) continue;
            long long fo = (fOne[k - 1] + 2 * fOne[k] + fOne[k + 1]) % MOD;
            ans = (ans + fz * fo) % MOD;
        }
        
        return static_cast<int>(ans);
    }
};

File: 3129-find-all-possible-stable-binary-arrays-i/3129-find-all-possible-stable-binary-arrays-i.py
class Solution:
    def numberOfStableArrays(self, zero: int, one: int, limit: int) -> int:
        MOD = 1000000007
        maxN = zero + one
        
        fact = [0] * (maxN + 1)
        invFact = [0] * (maxN + 1)
        
        fact[0] = 1
        invFact[0] = 1
        for i in range(1, maxN + 1):
            fact[i] = (fact[i - 1] * i) % MOD
            
        invFact[maxN] = pow(fact[maxN], MOD - 2, MOD)
        for i in range(maxN - 1, 0, -1):
            invFact[i] = (invFact[i + 1] * (i + 1)) % MOD
            
        def C(n, k):
            if k < 0 or k > n:
                return 0
            return fact[n] * invFact[k] % MOD * invFact[n - k] % MOD

        def F(N, K, L):
            if K <= 0 or K > N:
                return 0
            ans = 0
            maxJ = (N - K) // L
            for j in range(maxJ + 1):
                term = C(K, j) * C(N - j * L - 1, K - 1) % MOD
                if j & 1:
                    ans = (ans - term + MOD) % MOD
                else:
                    ans = (ans + term) % MOD
            return ans

        maxK = min(zero, one + 1)
        fOne = [0] * (maxK + 2)
        for k in range(1, maxK + 2):
            fOne[k] = F(one, k, limit)
            
        ans = 0
        for k in range(1, maxK + 1):
            fz = F(zero, k, limit)
            if fz == 0:
                continue
            fo = (fOne[k - 1] + 2 * fOne[k] + fOne[k + 1]) % MOD
            ans = (ans + fz * fo) % MOD
            
        return ans